py/desispec/io/emlinefit.py

```diff
@@ -0,0 +1,567 @@
+        h["{}_IVAR".format(camera)].data = h["{}_IVAR".format(camera)].data[ii_co, :]
+    # AR b : 3600,5800  , r : 5760,7620 , z : 7520,9824
+    # AR we keep all data points (i.e. including overlaps)
+    # AR note that waves is not ordered, as a consequence
+    waves = np.zeros(0)
+    fluxes = np.zeros(0).reshape(nspec, 0)
+    ivars = np.zeros(0).reshape(nspec, 0)
+    for camera in cameras:
+        tmpw = h["{}_WAVELENGTH".format(camera)].data
+        tmpfl = h["{}_FLUX".format(camera)].data
+        tmpiv = h["{}_IVAR".format(camera)].data
+        # AR TBD: use a smarter way with no loop on tids...
+        # AR TBD: but as 500 rows at most, ~ok
+        tmpexts = ext_odonnell(tmpw, Rv=rv)
+        for i in range(nspec):
+            tmp_mw_trans =  10 ** (-0.4 * ebvs[i] * rv * tmpexts)
+            tmpfl[i, :] /= tmp_mw_trans
+            tmpiv[i, :] *= tmp_mw_trans ** 2
+        waves = np.append(waves, tmpw)
+        fluxes = np.append(fluxes, tmpfl, axis=1)
+        ivars = np.append(ivars, tmpiv, axis=1)
+    # AR dictionary to store results
+    # AR keys returned by emlines_gaussfit
+    emkeys = [
+        "FLUX", "FLUX_IVAR", "SIGMA", "SIGMA_IVAR",
+        "CONT", "CONT_IVAR", "SHARE", "SHARE_IVAR",
+        "EW", "EW_IVAR",
+        "CHI2", "NDOF",
+    ]
+    mydict = {}
+    # AR adding rr_keys, fm_keys
+    for key in rr_keys.split(","):
+        if key == "TARGETID":
+            mydict["TARGETID"] = targetids
+        else:
+            mydict[key] = d_rr[key]
+    for key in fm_keys.split(","):
+        mydict[key] = d_fm[key]
+    # AR quantities required for the pdf if outpdf
+    if outpdf is not None:
+        spectypes, deltachi2s, objtypes = d_rr["SPECTYPE"], d_rr["DELTACHI2"], d_fm["OBJTYPE"]
+    # AR fit columns
+    for emname in emnames:
+        mydict[emname] = {}
+        for key in emkeys:
+            if key == "NDOF":
+                mydict[emname][key] = -99 + np.zeros(nspec, dtype=int)
+            else:
+                mydict[emname][key] = np.nan + np.zeros(nspec)
+        for key in ["wave", "fitdata", "fitivar", "model"]:
+            mydict[emname][key] = np.zeros(nspec, dtype=object)
+    for i_emname, emname in enumerate(emnames):
+        for i in range(nspec):
+            tmpmydict, _, w, d, iv, m = emlines_gaussfit(
+                emname,
+                targetids[i],
+                d_rr["Z"][i],
+                waves,
+                fluxes[i, :],
+                ivars[i, :],
+                rf_fit_hw=rf_fit_hw,
+                min_rf_fit_hw=min_rf_fit_hw,
+                rf_cont_w=rf_cont_w,
+                log=log,
+            )
+            for key in emkeys:
+                mydict[emname][key][i] = tmpmydict[key]
+            mydict[emname]["wave"][i] = w
+            mydict[emname]["fitdata"][i] = d
+            mydict[emname]["fitivar"][i] = iv
+            mydict[emname]["model"][i] = m
+    # AR plot?
+    if outpdf is not None:
```

Comment:
Usually we want to separate calculation from plotting, e.g. so that if you want to update the plots you can call it with some data to get a new plot without having to redo intermediate calculations.  Could you move out this plotting into a separate function that would take the outputs of `emlines_gaussfit` and make a plot, or does it rely upon some internal variable that isn't returned?

Reply:
done, I've moved that to `desispec.io.emlinefit.plot_emlines()`.